#*--------------- Extend ---------------*
extends Panel
class_name Tile

#*--------------- Extported Fields ---------------*
@export_group("Nodes")
@export var button: Button
@export var label: Label
@export var flag_label: Label

@export_group("Properties")
@export var mine_rate = .2

#*--------------- Fields ---------------*
var tile_position := Vector2i.ZERO
var mine_count := 0

## ã‚¯ãƒªã‚¢ãƒ•ãƒ©ã‚°
var is_cleared = false

## çˆ†å¼¾ãƒ•ãƒ©ã‚°
var is_mine = false

## ãƒ•ãƒ©ã‚°ãƒ•ãƒ©ã‚°
var is_flag = false

## ãƒœã‚¿ãƒ³æ“ä½œå¯èƒ½ãƒ•ãƒ©ã‚°
var button_operatable = true

#*--------------- Signals ---------------*
## ãƒœã‚¿ãƒ³ã‚¯ãƒªãƒƒã‚¯æ™‚ã‚·ã‚°ãƒŠãƒ«
signal on_tile_clicked

## çˆ†å¼¾ã‚¯ãƒªãƒƒã‚¯æ™‚ã‚·ã‚°ãƒŠãƒ«
signal on_mine_clicked

## ãƒ•ãƒ©ã‚°å¤‰æ›´æ™‚ã‚·ã‚°ãƒŠãƒ«
signal on_flag_changed(is_flag: bool)

#*--------------- Methods ---------------*
## ç”Ÿæˆå‡¦ç†
func generated(_position: Vector2i):
	tile_position = _position
	pass

## çˆ†å¼¾ã‚«ã‚¦ãƒ³ãƒˆã‚»ãƒƒãƒˆ
func set_mine_count(_count: int):
	if is_mine: return

	mine_count = _count

	# çˆ†å¼¾æ•°ãŒ0ä»¥ä¸Šã®å ´åˆã€ãƒ©ãƒ™ãƒ«ã«è¡¨ç¤º
	if mine_count > 0:
		label.text = str(_count)
	else:
		label.text = ""

## çˆ†å¼¾ã‚»ãƒƒãƒˆ
func set_mine():
	if is_cleared: return

	is_mine = true
	label.text = "ğŸ’£"

## ãƒ•ãƒ©ã‚°ã‚»ãƒƒãƒˆ
func set_flag():
	if is_cleared: return

	is_flag = !is_flag
	if is_flag:
		flag_label.text = "ğŸš©"
	else:
		flag_label.text = ""

	on_flag_changed.emit(is_flag)

## ãƒœã‚¿ãƒ³æ“ä½œå¯èƒ½ãƒ•ãƒ©ã‚°ã‚»ãƒƒãƒˆ
func set_button_operatable(_operatable: bool):
	button_operatable = _operatable

## ã‚¯ãƒªã‚¢å‡¦ç†
func cleared():
	if is_cleared or is_flag: return

	is_cleared = true
	button.visible = false

	# çˆ†å¼¾ã‚¯ãƒªãƒƒã‚¯æ™‚ã®å‡¦ç†
	if is_mine:
		on_mine_clicked.emit()
	
	on_tile_clicked.emit()

#------------------------------------------------------------

func _on_button_pressed() -> void:
	# æ“ä½œä¸å¯ã®å ´åˆã€å‡¦ç†ã—ãªã„
	if not button_operatable: return

	cleared()

# å³ã‚¯ãƒªãƒƒã‚¯åˆ¤å®šç”¨ã‚·ã‚°ãƒŠãƒ«
func _on_button_gui_input(event: InputEvent) -> void:
	# æ“ä½œä¸å¯ã®å ´åˆã€å‡¦ç†ã—ãªã„
	if not button_operatable: return

	if event is InputEventMouseButton:
		if event.button_index == MOUSE_BUTTON_RIGHT and event.pressed:
			set_flag()
